src/ApplicationInsights.Kubernetes/K8sHttpClient/KubeHttpClientSettingsBase.cs

```diff
@@ -0,0 +1,137 @@
+            if (string.IsNullOrEmpty(kubernetesServicePort))
+            {
+                throw new NullReferenceException("Kubernetes service port is not set.");
+            }
+
+            string baseAddress = Invariant($"https://{kubernetesServiceHost}:{kubernetesServicePort}/");
+            _logger.LogDebug(Invariant($"Kubernetes base address: {baseAddress}"));
+            ServiceBaseAddress = new Uri(baseAddress, UriKind.Absolute);
+        }
+
+        public string ContainerId { get; protected set; }
+
+        public string QueryNamespace { get; protected set; }
+
+        public Uri ServiceBaseAddress { get; private set; }
+
+        public virtual HttpMessageHandler CreateMessageHandler()
+        {
+            string certFilePath = GetCertFilePath();
+            if (!File.Exists(certFilePath))
+            {
+                throw new FileNotFoundException("Certificate file is required to access kubernetes API.", certFilePath);
+            }
+
+            HttpClientHandler handler = new HttpClientHandler()
+            {
+                ServerCertificateCustomValidationCallback = (httpRequestMessage, serverCert, chain, policyErrors) =>
+                {
+                    X509Certificate2 clientCert = new X509Certificate2(certFilePath);
+                    return CertificateValidationCallBack(httpRequestMessage, serverCert, clientCert, chain, policyErrors);
+                }
+            };
+
+            return handler;
+        }
+
+        /// <summary>
+        /// SSl Cert Validation Callback
+        /// </summary>
+        /// <param name="requestMessage">Http request message.</param>
+        /// <param name="caCert">Server certificate.</param>
+        /// <param name="clientCertificate">Client certificate</param>
+        /// <param name="chain">Chain</param>
+        /// <param name="sslPolicyErrors">SSL policy errors</param>
+        /// <returns>true if valid cert</returns>
+        internal bool CertificateValidationCallBack(
+#pragma warning disable CA1801 // Unused by design
+            HttpRequestMessage requestMessage,
+#pragma warning restore CA1801 // Restore the warning
+            X509Certificate2 caCert,
+            X509Certificate2 clientCertificate,
+            X509Chain chain,
+            SslPolicyErrors sslPolicyErrors)
+        {
+            // If the certificate is a valid, signed certificate, return true.
+            if (sslPolicyErrors == SslPolicyErrors.None)
+            {
+                _logger.LogTrace("This is a valid signed certificate.");
+                return true;
+            }
+
+            _logger.LogTrace("Not a authority signed certificate.");
+            _logger.LogTrace("Server Cert RAW: " + Environment.NewLine + Convert.ToBase64String(caCert.RawData));
+            // If there are errors in the certificate chain, look at each error to determine the cause.
```

Comment:
This is a comment taken from K8s client code, but it is misleading. The code below basically turns off certificate authority checking and certificate revocation checking, allowing certificates signed by unknown parties. In other words, as long as the certificate is not expired and it was issued for the right endpoint, it will be accepted regardless who issued it, or whether it was revoked.

Are you sure this is the right thing to do in this case?

If so, a better comment explaining the rationale would be very helpful here.

Reply:
If we are dealing with self-signed certificates here (self-signed by K8s cluster), maybe we can learn what the K8s cluster authority is and configure it explicitly as part of SSL checking policy (as a trusted authority). That would be **much better and stronger validation** than turning off cert authority check/cert revocation altogether.

Reply:
:-) Yes, we are dealing with Self-Signed Certification here. My understand: the point of having cert validation is to prevent MITM. We are using the local cert + the returned server cert to build the chain. The chain must be valid. Refer line 101:
```csharp
      bool isValid = chain.Build(clientCertificate); // the server cert is in the chain object already
```

Unit tests has been added to verify both success & fail scenarios:
```
public void TrueWhenValidCertificate()
public void FalseWhenInvalidCertificate()
```

Reply:
Comment updated.